lib/cli/src/commands/init.rs

```diff
@@ -0,0 +1,472 @@
+
+        if target_file.exists() && !self.overwrite {
+            return Err(anyhow::anyhow!(
+                "wapm project already initialized in {}",
+                target_file.display()
+            ));
+        }
+
+        // See if the directory has a Cargo.toml file, if yes, copy the license / readme, etc.
+        let manifest_path = match self.manifest_path.as_ref() {
+            Some(s) => s.clone(),
+            None => {
+                let cargo_toml_path = self
+                    .out
+                    .clone()
+                    .unwrap_or_else(|| std::env::current_exe().unwrap())
+                    .join("Cargo.toml");
+                cargo_toml_path
+                    .canonicalize()
+                    .unwrap_or_else(|_| cargo_toml_path.clone())
+            }
+        };
+
+        let cargo_toml = if manifest_path.exists() {
+            use anyhow::Context;
+
+            let mut metadata = MetadataCommand::new();
+            metadata.manifest_path(&manifest_path);
+            metadata.no_deps();
+            metadata.features(CargoOpt::AllFeatures);
+
+            println!("{:#?}", metadata.cargo_command());
+
+            let metadata = metadata.exec();
+
+            let metadata = match metadata {
+                Ok(o) => o,
+                Err(e) => {
+                    return Err(anyhow::anyhow!("failed to load metadata: {e}")
+                        .context(anyhow::anyhow!("{}", manifest_path.display())));
+                }
+            };
+
+            let package = metadata
+                .root_package()
+                .ok_or_else(|| anyhow::anyhow!("no root package found in cargo metadata"))
+                .context(anyhow::anyhow!("{}", manifest_path.display()))?;
+
+            Some(MiniCargoTomlPackage {
+                name: package.name.clone(),
+                version: package.version.clone(),
+                description: package.description.clone(),
+                homepage: package.homepage.clone(),
+                repository: package.repository.clone(),
+                license: package.license.clone(),
+                readme: package.readme.clone().map(|s| s.into_std_path_buf()),
+                license_file: package.license_file.clone().map(|f| f.into_std_path_buf()),
+                workspace_root: metadata.workspace_root.into_std_path_buf(),
+                build_dir: metadata
+                    .target_directory
+                    .into_std_path_buf()
+                    .join("wasm32-wasi"),
+            })
+        } else {
+            None
+        };
+
+        let package_name = cargo_toml
+            .as_ref()
+            .map(|p| &p.name)
+            .unwrap_or(&package_name);
+
+        let namespace = self.namespace.clone().unwrap_or_else(|| {
+            let username = wasmer_registry::whoami(None, None).ok().map(|o| o.1);
+            username
+                .or_else(|| package_name.split('/').next().map(|s| s.to_string()))
```

Comment:
This feels like a code smell. 

If the `package_name` came from the `cargo_toml` variable then it can never contain a `/`. Similarly, if you scroll up about 100 lines you'll see that `package_name` is constructed using *just* the `file_stem()` of the current directory. That means the `split()` will always return the entire `package_name` string and we don't need to deal with any `Option`s.

The deeper problem is that we shouldn't be using `string.split("/")` when dealing with package names because package names are more nuanced than just a string. Instead, we should try to use a high-level type and do something like `string.parse::<PackageName>()` so we can access its namespace.

Reply:
This is for cases where the name is constructed from the `wasmer.toml`, in the `init` step, I am currently setting the default name to `_` (since that user doesn't exist, we reserve it for commands), so it's a good default.

The `--namespace` option exists so I can change the username quickly.

Reply:
Continued in the follow-up comment: https://github.com/wasmerio/wasmer/pull/3324#discussion_r1049349754